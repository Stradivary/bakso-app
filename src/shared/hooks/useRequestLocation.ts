// useRequestLocation.ts
import { useState } from "react";

interface LocationState {
  latitude: number | null;
  longitude: number | null;
  error: string | null;
  isLoading: boolean;
}

export const useRequestLocation = () => {
  const [state, setState] = useState<LocationState>({
    latitude: null,
    longitude: null,
    error: null,
    isLoading: true,
  });

  const requestLocation = () => {
    if (!navigator.geolocation) {
      setState((prev) => ({
        ...prev,
        error: "Browser anda tidak mendukung geolocation",
        isLoading: false,
      }));
      return;
    }

    setState((prev) => ({ ...prev, isLoading: true }));

    navigator.geolocation.getCurrentPosition(
      (position) => {
        setState((prev) => ({
          ...prev,
          latitude: position.coords.latitude,
          longitude: position.coords.longitude,
          error: null,
          isLoading: false,
        }));
      },
      (error) => {
        setState({
          latitude: null,
          longitude: null,
          error: error.message,
          isLoading: false,
        });
      },
    );
  };

  return { ...state, requestLocation };
};
