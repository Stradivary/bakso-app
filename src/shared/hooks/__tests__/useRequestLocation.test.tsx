import { act, renderHook } from "@testing-library/react";
import {
  afterAll,
  afterEach,
  beforeAll,
  describe,
  expect,
  it,
  vi,
} from "vitest";
import { useRequestLocation } from "../useRequestLocation";

// Define mock coordinates type for better TypeScript support
interface MockCoordinates {
  latitude: number;
  longitude: number;
  accuracy: number;
  altitude: null | number;
  altitudeAccuracy: null | number;
  heading: null | number;
  speed: null | number;
}

// Default mock coordinates
const defaultCoords: MockCoordinates = {
  latitude: -6.2088,
  longitude: 106.8456,
  accuracy: 100,
  altitude: null,
  altitudeAccuracy: null,
  heading: null,
  speed: null,
};

// Create mock geolocation object
const mockGeolocation = {
  getCurrentPosition: vi.fn().mockImplementation((success) =>
    Promise.resolve(
      success({
        coords: defaultCoords,
        timestamp: Date.now(),
      }),
    ),
  ),
  watchPosition: vi.fn(),
  clearWatch: vi.fn(),
};

// Setup before tests
beforeAll(() => {
  vi.useFakeTimers();
  Object.defineProperty(global.navigator, "geolocation", {
    value: mockGeolocation,
    writable: true,
  });
});

// Clean up after tests
afterAll(() => {
  vi.unstubAllGlobals();
});

// Optional: Reset mocks between tests
afterEach(() => {
  vi.clearAllMocks();
});

describe("useRequestLocation", () => {
  it("should initialize with default values", () => {
    const { result } = renderHook(() => useRequestLocation());
    expect(result.current).toMatchSnapshot();
  });

  it("should handle successful location request", async () => {
    const mockPosition = {
      coords: {
        latitude: 40.7128,
        longitude: -74.006,
      },
    };

    global.navigator.geolocation.getCurrentPosition = vi.fn((success) =>
      success(mockPosition),
    );

    const { result } = renderHook(() => useRequestLocation());

    act(() => {
      result.current.requestLocation();
    });

    expect(result.current).toMatchSnapshot();
  });

  it("should handle location error", () => {
    const mockError = new Error("Location denied");
    global.navigator.geolocation.getCurrentPosition = vi.fn((_, error) =>
      error(mockError),
    );

    const { result } = renderHook(() => useRequestLocation());

    act(() => {
      result.current.requestLocation();
    });

    expect(result.current).toMatchSnapshot();
  });
});
