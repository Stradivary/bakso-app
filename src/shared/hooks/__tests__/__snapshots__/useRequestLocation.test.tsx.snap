// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`useRequestLocation > should handle location error 1`] = `
{
  "error": "Location denied",
  "isLoading": false,
  "latitude": null,
  "longitude": null,
  "requestLocation": [Function],
}
`;

exports[`useRequestLocation > should handle successful location request 1`] = `
{
  "error": null,
  "isLoading": false,
  "latitude": 40.7128,
  "longitude": -74.006,
  "requestLocation": [Function],
}
`;

exports[`useRequestLocation > should initialize with default values 1`] = `
{
  "error": null,
  "isLoading": true,
  "latitude": null,
  "longitude": null,
  "requestLocation": [Function],
}
`;
